File: templates/templates/javav2/tests/primary/GlobalsAdditionalTest.java
package org.openapis.openapi;

import org.junit.jupiter.api.Test;
import org.openapis.openapi.models.operations.*;

import static org.junit.jupiter.api.Assertions.*;
import static org.openapis.openapi.Helpers.HTTPBIN_URL;

public class GlobalsAdditionalTest {

    @Test
    void testGlobalsQueryParameterGetUsesGlobal() throws Exception {
        CommonHelpers.recordTest("globals-query-parameter-get-uses-global");

        SDK s = SDK.builder().serverURL(Helpers.HTTPBIN_URL).globalQueryParam("test").build();
        assertNotNull(s);

        GlobalsQueryParameterGetResponse res = s.globals().globalsQueryParameterGet().call();

        assertNotNull(res);
        assertEquals(200, res.statusCode());
        assertEquals("test", res.res().get().args().globalQueryParam());
    }

    @Test
    void testGlobalsQueryParameterGetUsesLocal() throws Exception {
        CommonHelpers.recordTest("globals-query-parameter-get-uses-local");

        SDK s = SDK.builder().serverURL(Helpers.HTTPBIN_URL).globalQueryParam("test").build();
        assertNotNull(s);

        GlobalsQueryParameterGetResponse res =
                s.globals().globalsQueryParameterGet().globalQueryParam("local").call();

        assertNotNull(res);
        assertEquals(200, res.statusCode());
        assertEquals("local", res.res().get().args().globalQueryParam());
    }

    @Test
    void testGlobalPathParameterGetUsesGlobal() throws Exception {
        CommonHelpers.recordTest("globals-path-parameter-get-uses-global");

        SDK s = SDK.builder().serverURL(Helpers.HTTPBIN_URL).globalPathParam(1L).build();
        assertNotNull(s);

        GlobalPathParameterGetResponse res = s.globals().globalPathParameterGet().call();

        assertNotNull(res);
        assertEquals(200, res.statusCode());
        assertEquals(HTTPBIN_URL + "/anything/globals/pathParameter/1",
                res.res().get().url());
    }

    @Test
    void testGlobalPathParameterGetUsesLocal() throws Exception {
        CommonHelpers.recordTest("globals-path-parameter-get-uses-local");

        SDK s = SDK.builder().serverURL(Helpers.HTTPBIN_URL).globalPathParam(1L).build();
        assertNotNull(s);

        GlobalPathParameterGetResponse res =
                s.globals().globalPathParameterGet().globalPathParam(2L).call();

        assertNotNull(res);
        assertEquals(200, res.statusCode());
        assertEquals(HTTPBIN_URL + "/anything/globals/pathParameter/2",
                res.res().get().url());
    }

    @Test
    void testGlobalHeaderGetUsesGlobal() throws Exception {
        CommonHelpers.recordTest("globals-header-get-uses-global");

        SDK s = SDK.builder().serverURL(Helpers.HTTPBIN_URL).globalHeaderParam(true).build();
        assertNotNull(s);

        GlobalsHeaderGetResponse res = s.globals().globalsHeaderGet().call();

        assertNotNull(res);
        assertEquals(200, res.statusCode());
        assertEquals("true", res.res().get().headers().get().get("Globalheaderparam"));
    }

    @Test
    void testGlobalHeaderGetUsesLocal() throws Exception {
        CommonHelpers.recordTest("globals-header-get-uses-local");

        SDK s = SDK.builder().serverURL(Helpers.HTTPBIN_URL).globalHeaderParam(true).build();
        assertNotNull(s);

        GlobalsHeaderGetResponse res =
                s.globals().globalsHeaderGet().globalHeaderParam(false).call();

        assertNotNull(res);
        assertEquals(200, res.statusCode());
        assertEquals("false", res.res().get().headers().get().get("Globalheaderparam"));
    }

    @Test
    void testGlobalsHiddenPost() throws Exception {
        CommonHelpers.recordTest("globals-hidden-post");

        SDK s = SDK.builder().serverURL(Helpers.HTTPBIN_URL)
                .globalHiddenQueryParam("hello")
                .globalHiddenHeaderParam("world")
                .globalHiddenPathParam("test")
                .build();

        var body = GlobalsHiddenPostRequestBody.builder()
                .test("friend")
                .other(37)
                .build();

        var res = s.globals().globalsHiddenPost(body);

        assertNotNull(res);
        assertEquals(200, res.statusCode());
        assertEquals("hello", res.res().get().args().globalHiddenQueryParam());
        assertEquals("friend", res.res().get().json().test());
        assertEquals(37, res.res().get().json().other());
        assertEquals("world", res.res().get().headers().get("Globalhiddenheaderparam"));
        assertEquals(HTTPBIN_URL + "/anything/globals/hidden/test?globalHiddenQueryParam=hello",
                res.res().get().url());
    }

    @Test
    void testGlobalsOperationParamsOnly() throws Exception {
        CommonHelpers.recordTest("globals-operation-params-only");

        // Initialize SDK with ALL global parameters
        SDK s = SDK.builder().serverURL(Helpers.HTTPBIN_URL)
                .globalQueryParam("globalQueryValue")
                .globalPathParam(999L)
                .globalHeaderParam(true)
                .globalHiddenQueryParam("hiddenQueryValue")
                .globalHiddenHeaderParam("hiddenHeaderValue")
                .globalHiddenPathParam("hiddenPathValue")
                .build();

        assertNotNull(s);

        // Call operation with operation-specific parameters
        GlobalsOperationScopedExclusiveResponse res = s.globals()
                .globalsOperationScopedExclusive()
                .operationQueryParam("operationQuery")
                .operationPathParam("operationPath")
                .operationHeaderParam("operationHeader")
                .call();

        assertNotNull(res);
        assertEquals(200, res.statusCode());
        assertTrue(res.res().isPresent());
        var payload = res.res().get();

        // Verify that ONLY operation-specific parameters are present in the request
        // Query params should NOT contain global params
        var args = payload.args();
        assertFalse(args.containsKey("globalQueryParam"),
                "Global query param should NOT be included when operation defines its own params");
        assertFalse(args.containsKey("globalHiddenQueryParam"),
                "Global hidden query param should NOT be included when operation defines its own params");

        // URL should contain operation path param, NOT global path param
        var url = payload.url();
        assertEquals(HTTPBIN_URL + "/anything/globals/operationScopedExclusive/operationPath?operationQueryParam=operationQuery", url,
                "URL should use operation-specific path parameter, not global");

        // Headers should NOT contain global headers
        var headers = payload.headers();
        assertFalse(headers.containsKey("Globalheaderparam"),
                "Global header param should NOT be included when operation defines its own params");
        assertFalse(headers.containsKey("Globalhiddenheaderparam"),
                "Global hidden header param should NOT be included when operation defines its own params");
    }

    @Test
    void testGlobalsKebabCaseParamGet() throws Exception {
        CommonHelpers.recordTest("globals-kebab-case-param-get");

        SDK s = SDK.builder().serverURL(Helpers.HTTPBIN_URL).kebabCaseParam("kebab-case-value").build();
        assertNotNull(s);

        GlobalsKebabCaseParamGetResponse res = s.globals().globalsKebabCaseParamGet().call();

        assertNotNull(res);
        assertEquals(200, res.statusCode());
        assertTrue(res.res().isPresent());
        assertEquals("kebab-case-value", res.res().get().args().get("kebab-case-param"));
    }
}

File: templates/templates/typescriptv2/tests/no-zod/globals_additional.test.ts
import { vi, expect, test, afterEach } from "vitest";

import { SDK } from "../index.js";
import { resetEnv } from "../lib/env.js";
import { HTTPClient } from "../lib/http.js";
import { GlobalPathParameterGetResponse } from "../sdk/models/operations/globalpathparameterget.js";
import { GlobalsHeaderGetResponse } from "../sdk/models/operations/globalsheaderget.js";
import { GlobalsHiddenPostResponse } from "../sdk/models/operations/globalshiddenpost.js";
import { GlobalsQueryParameterGetResponse } from "../sdk/models/operations/globalsqueryparameterget.js";

import { recordTest, HTTPBIN_URL } from "./common_helpers.js";

afterEach(() => {
  resetEnv();
  vi.unstubAllEnvs();
});

test("Test Globals Query Parameter Get Uses Global", async () => {
  recordTest("globals-query-parameter-get-uses-global");

  vi.stubEnv("SPEAKEASY_GLOBAL_QUERY_PARAM", "test");

  const sdk = new SDK({ serverURL: HTTPBIN_URL });

  const res: GlobalsQueryParameterGetResponse =
    await sdk.globals.globalsQueryParameterGet();

  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res?.args.globalQueryParam).toBe("test");
});

test("Test Globals Query Parameter Get Uses Local", async () => {
  recordTest("globals-query-parameter-get-uses-local");

  const sdk = new SDK({
    serverURL: HTTPBIN_URL,
    globalQueryParam: "test",
  });

  const res: GlobalsQueryParameterGetResponse =
    await sdk.globals.globalsQueryParameterGet("local");

  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res?.args.globalQueryParam).toBe("local");
});

test("Test Globals Path Parameter Get Uses Global", async () => {
  recordTest("globals-path-parameter-get-uses-global");

  const sdk = new SDK({
    serverURL: HTTPBIN_URL,
    globalPathParam: 1,
  });

  const res: GlobalPathParameterGetResponse =
    await sdk.globals.globalPathParameterGet();

  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res?.url).toBe(`${HTTPBIN_URL}/anything/globals/pathParameter/1`);
});

test("Test Globals Path Parameter Get Uses Local", async () => {
  recordTest("globals-path-parameter-get-uses-local");

  const sdk = new SDK({
    serverURL: HTTPBIN_URL,
    globalPathParam: 1,
  });

  const res: GlobalPathParameterGetResponse =
    await sdk.globals.globalPathParameterGet(2);

  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res?.url).toBe(`${HTTPBIN_URL}/anything/globals/pathParameter/2`);
});

test("Test Globals Header Get Uses Global", async () => {
  recordTest("globals-header-get-uses-global");

  const sdk = new SDK({
    serverURL: HTTPBIN_URL,
    globalHeaderParam: true,
  });

  const res: GlobalsHeaderGetResponse = await sdk.globals.globalsHeaderGet();

  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res?.headers?.["Globalheaderparam"]).toBe("true");
});

test("Test Globals Header Get Uses Local", async () => {
  recordTest("globals-header-get-uses-local");

  const sdk = new SDK({
    serverURL: HTTPBIN_URL,
    globalHeaderParam: true,
  });

  const res: GlobalsHeaderGetResponse =
    await sdk.globals.globalsHeaderGet(false);

  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res?.headers?.["Globalheaderparam"]).toBe("false");
});

test("Test Global Header Keeps Custom Client Headers", async () => {
  recordTest("globals-header-keeps-custom-client-headers");

  const httpClient = new HTTPClient({
    fetcher: (request) => {
      return fetch(request);
    },
  });

  httpClient.addHook("beforeRequest", (request) => {
    request.headers.set("x-custom-header", "customValue");

    return request;
  });

  const sdk = new SDK({
    serverURL: HTTPBIN_URL,
    globalHeaderParam: true,
    httpClient: httpClient,
  });

  const res: GlobalsHeaderGetResponse = await sdk.globals.globalsHeaderGet();

  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res?.headers?.["Globalheaderparam"]).toBe("true");
  expect(res.res?.headers?.["X-Custom-Header"]).toBe("customValue");
});

test("Test Globals Hidden Post", async () => {
  recordTest("globals-hidden-post");

  const sdk = new SDK({
    serverURL: HTTPBIN_URL,
    globalHiddenQueryParam: "hello",
    globalHiddenHeaderParam: "world",
    globalHiddenPathParam: "test",
  });

  const res: GlobalsHiddenPostResponse = await sdk.globals.globalsHiddenPost({
    test: "friend",
    other: 37,
  });

  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res?.args.globalHiddenQueryParam).toBe("hello");
  expect(res.res?.json?.test).toBe("friend");
  expect(res.res?.json?.other).toBe(37);
  expect(res.res?.headers?.["Globalhiddenheaderparam"]).toBe("world");
  expect(res.res?.url).toBe(
    `${HTTPBIN_URL}/anything/globals/hidden/test?globalHiddenQueryParam=hello`,
  );
});

test("Test Globals Operation Params Only", async () => {
  recordTest("globals-operation-params-only");

  // Initialize SDK with ALL global parameters
  const sdk = new SDK({
    serverURL: HTTPBIN_URL,
    globalQueryParam: "globalQueryValue",
    globalPathParam: 999,
    globalHeaderParam: true,
    globalHiddenQueryParam: "hiddenQueryValue",
    globalHiddenHeaderParam: "hiddenHeaderValue",
    globalHiddenPathParam: "hiddenPathValue",
  });

  // Call operation with operation-specific parameters
  const res = await sdk.globals.globalsOperationScopedExclusive(
    "operationPath",
    "operationQuery",
    "operationHeader",
  );

  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res).toBeDefined();

  // Verify that ONLY operation-specific parameters are present in the request
  // Query params should NOT contain global params
  expect(res.res?.args["globalQueryParam"]).toBeUndefined();
  expect(res.res?.args["globalHiddenQueryParam"]).toBeUndefined();

  // URL should contain operation path param, NOT global path param
  expect(res.res?.url).toBe(
    `${HTTPBIN_URL}/anything/globals/operationScopedExclusive/operationPath?operationQueryParam=operationQuery`,
  );

  // Headers should NOT contain global headers
  expect(res.res?.headers?.["Globalheaderparam"]).toBeUndefined();
  expect(res.res?.headers?.["Globalhiddenheaderparam"]).toBeUndefined();
});

test("Test Globals Kebab Case Param Get", async () => {
  recordTest("globals-kebab-case-param-get");

  const sdk = new SDK({
    serverURL: HTTPBIN_URL,
    kebabCaseParam: "kebab-case-value",
  });

  const res = await sdk.globals.globalsKebabCaseParamGet();

  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res?.args["kebab-case-param"]).toBe("kebab-case-value");
});
